site/_posts/2020-11-15-how-to-securely-implement-oauth-rails.md

```diff
@@ -0,0 +1,262 @@
+
+We receive the access token encoded as a `JWT`. Finally, we save the JWT on the user session and redirect to the Welcome page. 
+
+The `logout` callback is much simpler. Similar to `oauth_callback` in the way that it receives the authorize redirect, our `destroy` method receives the logout redirect.
+Receiving the redirect tells us that the user has been logged out of FusionAuth and we are safe to clear the user's session in Rails.
+
+### Polishing up our Rails session
+Now that we have the functionality to authenticate users as well as log them out of FusionAuth, we will want to complete the same cycle on our Rails app.
+
+There are numerous ways to handle user sessions in Rails, but for this example, I decided to create helper methods accessible to all sub-classes of ApplicationController.
+
+`current_user` looks for a user jwt on the session. If it exists, it decodes the JWT and retrieves user's email address.
+If it does not exist, this means the user has logged out or their session has expired.
+
+```ruby
+# app/controllers/application_controller.rb
+
+class ApplicationController < ActionController::Base
+  helper_method :current_user
+  helper_method :logged_in?
+
+  def current_user
+    if session[:user_jwt]
+      token = session[:user_jwt]
+      decoded = TokenService.decode(token["value"])
+      decoded.first["email"]
+    end
+
+  end
+
+  def logged_in?
+    current_user.present?
+  end
+
+end
+```
+
+```ruby
+# app/services/token_service.rb
+
+class TokenService
+  def self.decode(token)
+    JWT.decode(
+        token,
+        nil,
+        false,
+        {algorithm: 'HS256'})
+  end
+end
+```
+
+## Putting it all together
+Before running the Rails server, make sure to add the following dependencies to your Gemfile.
+```
+gem 'jwt'
+gem 'rest-client'
+```
+Install dependencies
+```
+bundle install
+```
+
+Kick the tires and light the fires
+```
+rails s
+```
+
+Navigating to `http://localhost:3000`
+
+{% include _image.liquid src="/assets/img/blogs/fusionauth-example-rails/home.png" alt="Home" class="img-fluid" figure=false %}
+
+Login to the Application
+
+{% include _image.liquid src="/assets/img/blogs/fusionauth-example-rails/fa-login.png" alt="FusionAuth Login" class="img-fluid" figure=false %}
+
+We made it!
+
+{% include _image.liquid src="/assets/img/blogs/fusionauth-example-rails/logged-in.png" alt="Logged In" class="img-fluid" figure=false %}
+
+All done. Time to logout.
+
+{% include _image.liquid src="/assets/img/blogs/fusionauth-example-rails/logging-out.png" alt="Logging Out" class="img-fluid" figure=false %}
+
+## Next Steps
+Self service registration? FusionAuth can do that. MFA? Yes indeed! By integrating with FusionAuth, we can continue to enhance identity management for our users. 
```

Comment:
I might go lighter on the fusionauth cheerleading :) and suggest a couple of ways people could extend this rails app.

* give roles to a user, show and hide functionality based on that role
* customize the fusionauth theme and the view to have the same logo
* add a 'login with google link' and show the user that
* add self registration and link out to that if the user doesn't have a login
* ... other things.